### PR TITLE
Tally top 80 results exactly and fix "monday" timestamp calculation

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Fixes

Staci and I noticed two issues while we were deploying the API today:

1. If the page_size changes (for example to 200), then if we tally the first 4 pages, we'd be tallying the top 800 results, which does not really reflect the intention of the tallying (for seeing the provider density of "top results"
2. On weeks that include a month boundary, the "monday" date stamp calculation was incorrect. We only found this because we were working on this literally as 00:00 UTC 1 Feb 2023 rolled over from January!

## Description

Applies fixes for the above described issues. There are (hopefully) sufficient comments to hopefully describe the approach and intentions as the logic is a bit convoluted.

## Testing Instructions

Check out the much expanded unit tests. You can test locally following the process Staci describes here: https://github.com/WordPress/openverse-api/pull/1088#pullrequestreview-1266702601. However, because local results for any given query do not really have 80 results, you'll need to test without a query or with a `*` query.

## Checklist

- [x] My pull request has a descriptive title (not a vague title like
      `Update index.md`).
- [x] My pull request targets the _default_ branch of the repository (`main`) or
      a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible
      errors.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

### CODE REVIEW

File: api/catalog/api/controllers/search_controller.py [hunk 1]
```diff
@@ -414,10 +414,33 @@ def search(
...
-    if results and page <= 4:
-        # We ignore tallies for deep pages because they're not likely to
+    max_result_depth = page * page_size
+    if max_result_depth <= 80:
+        # Applies when `page_size * page` could land "evenly" on 80
+        should_tally = True
+        results_to_tally = results
+    elif max_result_depth - page_size < 80:
+        # Applies when `page_size * page` could land beyond 80, but still
+        # encompas on _this page_ some results that are below the 80th
```
Comment (AetherUnbound) [inline api/catalog/api/controllers/search_controller.py:424]:
```suggestion
        # encompass on _this page_ some results that are below the 80th
```

**Review by AetherUnbound — COMMENTED** [review]
Wow, nice find! It looks like there's a failing test, I'm able to produce it locally too

**Review by stacimc — COMMENTED** [review]
The test failure is reproducible for me as well :/ 

Testing manually the tallies LGTM, though. I tested:
* "normal" flow by not adjusting `page_size` and checking the tallies increase correctly after requesting page 1-4 but not after page 5
* using `page_size=7` and checked the tallies increase only for the first 80 values
* using `page_size =100` (after updating `MAX_ANONYMOUS_PAGE_SIZE` locally 😄) and verified only first 80 on page 1 are counted

**Comment by sarayourfriend** [comment]
Thanks for the reviews. The failing test was due to the fact that I'd accidentally removed the falsey check on `results`, so `None` was getting passed to the tallies function. It is fixed now (thank goodness for tests :sweat_smile:)

**Review by obulat — APPROVED** [review]
Perfect timing for deploy :)
LGTM!

**Review by dhruvkb — APPROVED** [review]
Interesting problem and solution, thanks @sarayourfriend!

**Pull request merged**